### PR TITLE
[Dist][Test] Add tests for multi-node dist dataloaders.

### PR DESCRIPTION
**Pull request by Kh4L**
Add tests for DistDataLoader, DistNodeDataLoader and DistEdgeDataLoader in a multi-node environment.

This PR implements #4376

### CODE REVIEW

**Review by Rhett-Ying — COMMENTED** [review]
is below test case covered? https://github.com/dmlc/dgl/blob/6566c31fbac4382c01125b56ce5ccc3ba1c50f87/tests/distributed/test_mp_dataloader.py#L712-L714

**Comment by Kh4L** [comment]
> is below test case covered?
> 
> https://github.com/dmlc/dgl/blob/6566c31fbac4382c01125b56ce5ccc3ba1c50f87/tests/distributed/test_mp_dataloader.py#L712-L714

Added a test covering this case

File: tests/dist/python/run_dist_objects.py [hunk 1]
```diff
@@ -392,6 +393,248 @@ def dist_optimizer_check_store(g):
...
+
+    dataloader = dgl.dataloading.DistDataLoader(
+        dataset=train_nid.numpy(),
+        batch_size=batch_size,
+        collate_fn=sampler.sample_blocks,
+        shuffle=shuffle,
+        drop_last=drop_last,
+    )
+
+    for epoch in range(2):
```
Comment (Rhett-Ying) [inline tests/dist/python/run_dist_objects.py:456]: is `epoch` used below? or just replace with `_`?

File: tests/dist/python/run_dist_objects.py [hunk 2]
```diff
@@ -392,6 +393,248 @@ def dist_optimizer_check_store(g):
...
+                    )
+                    assert np.all(
+                        F.asnumpy(block.dstnodes[dst_type].data[dgl.NID])
+                        == F.asnumpy(neg_graph.nodes[dst_type].data[dgl.NID])
+                    )
+                    assert pos_graph.num_edges() * num_negs == neg_graph.num_edges()
+    del dataloader
+
+def multi_distdataloader_test(g, dataloader_class):
+    num_items_to_sample=404
```
Comment (Rhett-Ying) [inline tests/dist/python/run_dist_objects.py:580]: could we remove magic numbers? how about use `node_split` to obtain `train_ids`?

File: tests/dist/python/run_dist_objects.py [hunk 3]
```diff
@@ -392,6 +393,248 @@ def dist_optimizer_check_store(g):
...
+    for ntype in g.ntypes:
+        orig_nid = F.tensor(np.load(graph_path + f'/orig_nid_array_{ntype}.npy'), dtype)
+        orig_nid_map[ntype] = orig_nid
+
+    orig_uv_map = dict()
+    for etype in g.etypes:
+        orig_uv_map[etype] = (g.edges[etype].data['edge_u'], g.edges[etype].data['edge_v'])
+
+    batch_size_l = [64]
+    num_items_to_sample_l = [202, 404]
```
Comment (Rhett-Ying) [inline tests/dist/python/run_dist_objects.py:620]: avoid magic numbers

File: tests/dist/python/run_dist_objects.py [hunk 4]
```diff
@@ -392,6 +393,248 @@ def dist_optimizer_check_store(g):
...
+
+def distnodedataloader_test(g, batch_size, num_nodes_to_sample, drop_last, shuffle,
+                            num_workers, orig_nid_map, orig_uv_map):
+    if len(g.ntypes) == 1:
+        train_nid = F.arange(0, num_nodes_to_sample)
+    else:
+        train_nid = {g.ntypes[-1]: F.arange(0, num_nodes_to_sample)}
+    sampler = dgl.dataloading.MultiLayerNeighborSampler(
+        [
+            {etype: 5 for etype in g.etypes}
```
Comment (Rhett-Ying) [inline tests/dist/python/run_dist_objects.py:492]: this snippet is not straightforward, could we refactor it. use explicit `if else` instead.

**Review by Rhett-Ying — APPROVED** [review]
generally looks good to me except several minor changes.

File: tests/dist/python/run_dist_objects.py (hunk 2 above)
Comment (Kh4L) [reply tests/dist/python/run_dist_objects.py:580]: This an arbitrary number of relatively small size to make the test  runtime shorter.

Running it with much bigger training set will be make the test very long (the dist dataloader loaders runtime is already 10-15min with 2 nodes with those numbers)

**Review by Rhett-Ying — APPROVED** [review]

**Pull request merged**